### PR TITLE
Stress test for RocksDB transactions

### PR DESCRIPTION
**Pull request by riversand963**
Current db_stress does not cover complex read-write transactions. Therefore, this PR adds
coverage for emulated MyRocks-style transactions in `MultiOpsTxnsStressTest`. To achieve this, we need:

- Add a new operation type 'customops' so that we can add new complex groups of operations, e.g. transactions involving multiple read-write operations.
- Implement three read-write transactions and two read-only ones to emulate MyRocks-style transactions.

Test plan:
```
make check
./db_stress -test_multi_ops_txns -use_txn -clear_column_family_one_in=0 -column_families=1 -writepercent=0 -delpercent=0 -delrangepercent=0 -customopspercent=60 -readpercent=20 -prefixpercent=0 -iterpercent=20 -reopen=0 -ops_per_thread=100000
```

Next step is to add more configurability and refine input generation and result reporting, which will done in separate follow-up PRs.

### CODE REVIEW

**Comment by riversand963** [comment]
cc @lth

File: db_stress_tool/myrocks_style_txns_stress.cc [hunk 1]
```diff
@@ -0,0 +1,1275 @@
...
+  ReadOptions ropts;
+  std::string value;
+  s = txn->GetForUpdate(ropts, old_pk, &value);
+  if (!s.ok()) {
+    return s;
+  }
+  std::string empty_value;
+  s = txn->GetForUpdate(ropts, new_pk, &empty_value);
+  if (s.ok()) {
+    s = Status::Busy();
```
Comment (lth) [inline db_stress_tool/myrocks_style_txns_stress.cc:793]: maybe assert `empty_value` is non-empty?

File: db_stress_tool/myrocks_style_txns_stress.cc [hunk 2]
```diff
@@ -0,0 +1,1275 @@
...
+
+  auto result = MyRocksRecord::DecodePrimaryIndexValue(value);
+  s = std::get<0>(result);
+  if (!s.ok()) {
+    return s;
+  }
+  uint32_t b = std::get<1>(result);
+  uint32_t c = std::get<2>(result);
+
+  s = txn->Delete(old_pk);
```
Comment (lth) [inline db_stress_tool/myrocks_style_txns_stress.cc:805]: In general, we call Put/Delete with `assume_tracked` on, on the assumption that we always read+lock before modifying (ie. no blind writes).

File: db_stress_tool/myrocks_style_txns_stress.cc [hunk 3]
```diff
@@ -0,0 +1,1275 @@
...
+    ropts.snapshot = txn->GetSnapshot();
+  }
+  ropts.total_order_seek = true;
+  ropts.iterate_upper_bound = &iter_ub;
+  it = txn->GetIterator(ropts);
+  it->Seek(old_sk_prefix);
+  if (!it->Valid()) {
+    s = Status::NotFound();
+    return s;
+  }
```
Comment (lth) [inline db_stress_tool/myrocks_style_txns_stress.cc:905]: I don't know if it's important to report NotFound to the upper layers, but the code can be simplified slightly if we just return OK if no rows matched for SK.

File: db_stress_tool/myrocks_style_txns_stress.cc [hunk 4]
```diff
@@ -0,0 +1,1275 @@
...
+                          s);
+        return;
+      }
+      // After decoding secondary index entry, we know a and c. Crc is verified
+      // in decoding phase.
+      //
+      // Form a primary key and search in the primary index.
+      std::string pk = MyRocksRecord::EncodePrimaryKey(record.a_value());
+      std::string value;
+      s = db_->Get(ReadOptions(), pk, &value);
```
Comment (lth) [inline db_stress_tool/myrocks_style_txns_stress.cc:1131]: Should this read be done using the appropriate snapshot?

File: db_stress_tool/myrocks_style_txns_stress.cc [hunk 5]
```diff
@@ -0,0 +1,1275 @@
...
+class MyRocksRecord {
+ public:
+  static constexpr uint32_t kPrimaryIndexId = 1;
+  static constexpr uint32_t kSecondaryIndexId = 2;
+
+  static constexpr size_t kPrimaryIndexEntrySize = 8 + 8;
+  static constexpr size_t kSecondaryIndexEntrySize = 12 + 4;
+
+  static_assert(kPrimaryIndexId < kSecondaryIndexId,
+                "kPrimaryIndexId must be smaller than kSecondaryIndexId");
```
[the remark this answers is not in the document]
Comment (riversand963) [reply db_stress_tool/myrocks_style_txns_stress.cc:131]: Right, this is a check for the current impl. We can remove it in the future.

File: db_stress_tool/myrocks_style_txns_stress.cc [hunk 6]
```diff
@@ -0,0 +1,1275 @@
...
+  char buf[8];
+  EncodeFixed32(buf, kPrimaryIndexId);
+  std::reverse(buf, buf + 4);
+  EncodeFixed32(buf + 4, a_);
+  std::reverse(buf + 4, buf + 8);
+  primary_index_key.assign(buf, sizeof(buf));
+
+  std::string primary_index_value;
+  EncodeFixed32(buf, b_);
+  std::reverse(buf, buf + 4);
```
[the remark this answers is not in the document]
Comment (riversand963) [reply db_stress_tool/myrocks_style_txns_stress.cc:387]: Will update.

File: db_stress_tool/myrocks_style_txns_stress.cc [hunk 7]
```diff
@@ -0,0 +1,1322 @@
...
+  primary_index_key.assign(buf, sizeof(buf));
+
+  std::string primary_index_value;
+  EncodeFixed32(buf, b_);
+  EncodeFixed32(buf + 4, c_);
+  primary_index_value.assign(buf, sizeof(buf));
+  return std::make_pair(primary_index_key, primary_index_value);
+}
+
+std::string MyRocksRecord::EncodePrimaryKey() const {
```
Comment (ajkr) [inline db_stress_tool/myrocks_style_txns_stress.cc:389]: Can it call the static one with `a_` as parameter?

File: db_stress_tool/myrocks_style_txns_stress.cc [hunk 8]
```diff
@@ -0,0 +1,1322 @@
...
+    return std::make_pair(Status::Corruption(""), 0);
+  }
+  uint32_t crc = 0;
+  bool result __attribute__((unused)) =
+      GetFixed32(&secondary_index_value, &crc);
+  assert(result);
+  return std::make_pair(Status::OK(), crc);
+}
+
+std::pair<std::string, std::string> MyRocksRecord::EncodePrimaryIndexEntry()
```
Comment (ajkr) [inline db_stress_tool/myrocks_style_txns_stress.cc:372]: Can it use EncodePrimaryKey and EncodePrimaryIndexValue?

File: db_stress_tool/myrocks_style_txns_stress.cc [hunk 9]
```diff
@@ -0,0 +1,1322 @@
...
+}
+
+Status MyRocksStyleTxnsStressTest::TestCheckpoint(
+    ThreadState* thread, const std::vector<int>& rand_column_families,
+    const std::vector<int64_t>& /*rand_keys*/) {
+  // TODO (yanqin)
+  (void)thread;
+  (void)rand_column_families;
+  return Status::OK();
+}
```
Comment (ajkr) [inline db_stress_tool/myrocks_style_txns_stress.cc:682]: It's interesting many of these need to be repeated across different `StressTest`s. I will think more about this later...

**Review by ajkr — COMMENTED** [review]
This is a very interesting and, I think, valuable test! Thanks!

Ignoring practical constraints (which I think don't offer us this choice right now), do you think such a test would belong in MyRocks repo? Assuming db_stress were extensible externally, CI setup were not an issue, etc.

Another question I had is, do you see any way to integrate with ExpectedState? For example, let the stress test think keys are `a * kInitNumC + c` and values are `b`. But then translate them internally to `a -> (b, c)` and `(c, a) -> b`. There would still be a big issue, which is locking.

**Comment by riversand963** [comment]
Thanks @ajkr for the review!

> Ignoring practical constraints (which I think don't offer us this choice right now), do you think such a test would belong in MyRocks repo? Assuming db_stress were extensible externally, CI setup were not an issue, etc.

This is a good point, and I gave it some thought which I'd be happy to discuss. :)
Generally, I think unit and stress tests for a specific user of RocksDB should **not** reside in RocksDB main repo. If we do this for one user/workload, then we may find it difficult to push back future requests to add workload-specific tests to RocksDB repo. If we cannot push back, then RocksDB repo may potentially grow with tests that others do not care about or are willing to maintain.
Ideally, if the restrictions mentioned above are gone (db_stress externally extensible, CI setup ready), it will be nice to put such workload-specific tests in a separate location, depending on who is willing to maintain it. Currently, if we put it in MyRocks repo, then we need to make sure that the stress tests are run daily on top of the latest commit in RocksDB. Arguably, running the tests infrequently may not be sufficient to discover issues in a timely manner.
I think we can make an exception for MyRocks this time, mainly due to restrictions mentioned above. Furthermore, I am confident that we won't have difficulty in finding people who are willing to maintain these tests, so we do not have to worry about it becoming obsolete and lying around without any attention. Also, we can see if we can remove the restrictions. Then, we can put the tests in a separate repo.

**Comment by riversand963** [comment]
> I assume we are testing RocksDB functionalities, not MySQL. It makes sense for the test code to stay here. Maybe we should remove those "MySQL" from class name or variable name. It might be confusing.

We can update the naming, but the data model and combination of operations are pretty specific to MyRocks. Maybe we can replace "MyRocks" with another word. Removing it, i.e. `TransactionStressTest`, etc are too general. @siying

**Comment by riversand963** [comment]
@ajkr 

> Another question I had is, do you see any way to integrate with ExpectedState

Did not dig too deep into this. I also think locking is a big challenge for expected state. Maybe use have an implementation of `ExpectedState` with locking support, e.g. another storage engine? Seems a little heavy-weight.
We currently rely on some pk/sk constraint and checksuming for data verification, which I guess, can cover a number of correctness tests.

**Comment by siying** [comment]
> @ajkr
> 
> > Another question I had is, do you see any way to integrate with ExpectedState
> 
> Did not dig too deep into this. I also think locking is a big challenge for expected state. Maybe use have an implementation of `ExpectedState` with locking support, e.g. another storage engine? Seems a little heavy-weight. We currently rely on some pk/sk constraint and checksuming for data verification, which I guess, can cover a number of correctness tests.

We are assuming
1. primary key is encoded as |index_id|col1|col2| and value is other values
2. secondary key is |index_id'|col1'|col2'|primary_keys|
Right?

I don't think it's very specific to MySQL, and it's a very straight-forward way of encoding without a lot of alternatives. Even if it is exactly the same as MyRocks now, it can be served as a general RocksDB test. If MyRocks change something, we don't need to change.

Regarding naming, we can name as something like ColumnEncodedTransactionStress or something like that?

File: db_stress_tool/multi_ops_txns_stress.h [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+// Copyright (c) 2011 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+#ifdef GFLAGS
+#include "db_stress_tool/db_stress_common.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+// This file defines MultiOpsTxnsStress so that we can stress test RocksDB
```
Comment (zhichao-cao) [inline db_stress_tool/multi_ops_txns_stress.h:15]: This part is cool. Very understandable

**Pull request closed without merging**